### PR TITLE
CLI Improvements

### PR DESCRIPTION
**Pull request by satwikkansal**
~~Requires https://github.com/coala/coala/pull/4622 to be merged first~~

^^ relevant commit splitted to #165

### CODE REVIEW

File: coala_quickstart/generation/SettingsFilling.py [hunk 1]
```diff
@@ -32,38 +35,41 @@ def fill_section(section,
...
-            needed_settings[setting] = help_text
+            needed_settings[setting] = setting_info
 
     # Fill the settings with existing values if possible
     satisfied_settings = []
 
     for setting in needed_settings.keys():
         to_fill_values = list(autofill_value_if_possible(
-            setting, section.name, needed_settings[setting][1:],
+            setting, section.name, needed_settings[setting]["bears"],
```
Comment (Nosferatul) [inline coala_quickstart/generation/SettingsFilling.py]: if you use "bears" a lot like in this, maybe you should define a global variable?

File: coala_quickstart/generation/SettingsFilling.py [hunk 2]
```diff
@@ -32,38 +35,41 @@ def fill_section(section,
...
         needed = bear.get_non_optional_settings()
         for key in needed:
             if key in prel_needed_settings:
-                prel_needed_settings[key].append(bear.name)
+                prel_needed_settings[key]["bears"].append(bear.name)
             else:
-                prel_needed_settings[key] = [needed[key][0],
-                                             bear.name]
+                prel_needed_settings[key] = {
+                    "help_text": needed[key][0],
```
Comment (Nosferatul) [inline coala_quickstart/generation/SettingsFilling.py]: I don't like 0, make a variable with a good name. I think it's more suggestive like that.

File: coala_quickstart/generation/SettingsFilling.py [hunk 3]
```diff
@@ -165,3 +171,109 @@ def resolve_anomaly(setting_name,
...
+    :param setting_info: A dictionary of the form
+
+    ::
+        {
+            "help_text": "help string for the setting",
+            "bears": [SomeBear", "SomeOtherBear"],
+            "type": bool
+        }
+
+    :param section:      The section the action corresponds to.
```
Comment (Nosferatul) [inline coala_quickstart/generation/SettingsFilling.py]: this should return something? if yes, please add doc for it.

**Comment by Nosferatul** [comment]
i rebuilded your pr, dumb error i think

File: coala_quickstart/generation/SettingsFilling.py (hunk 2 above)
Comment (satwikkansal) [inline coala_quickstart/generation/SettingsFilling.py]: Agreed. But this is based on existing design which I followed and definitely a topic for a separate issue/PR.

File: coala_quickstart/generation/SettingsFilling.py [hunk 4]
```diff
@@ -165,3 +168,93 @@ def resolve_anomaly(setting_name,
...
+    :return:             The preferred value provided by the user
+                         for the setting.
+    """
+    needed = join_names(setting_info["bears"])
+
+    STR_GET_VAL_FOR_SETTING = ('\nPlease enter a value for the setting \"{}\" '
+                               '({}) needed by {} for section \"{}\": ')
```
Comment (adtac) [inline coala_quickstart/generation/SettingsFilling.py]: Instead of `"{}"`, use `{}` and `repr(...)` the var you're using there when you `.format` it

Comment (adtac) [inline coala_quickstart/generation/SettingsFilling.py]: more robust IMO

File: coala_quickstart/generation/SettingsFilling.py [hunk 5]
```diff
@@ -165,3 +168,93 @@ def resolve_anomaly(setting_name,
...
+
+def require_setting(setting_name, setting_info, section):
+    """
+    This method is responsible for prompting a user about a missing setting and
+    taking its value as input from the user.
+
+    :param setting_name: Name of the setting missing
+    :param setting_info: A dictionary of the form
+
+    ::
```
Comment (adtac) [inline coala_quickstart/generation/SettingsFilling.py]: huh, I didn't know we did codeblocks this way - there's a bunch of docs that's strewn all over the code base without this

oh well, now's a good time to start as any

File: tests/generation/SettingsFillingTest.py [hunk 6]
```diff
@@ -7,12 +7,13 @@
 from coalib.bears.LocalBear import LocalBear
 from coala_utils.ContextManagers import (
     simulate_console_inputs, retrieve_stdout)
-from coalib.output.ConsoleInteraction import acquire_settings
+# from coalib.output.ConsoleInteraction import acquire_settings
```
Comment (adtac) [inline tests/generation/SettingsFillingTest.py]: meant to remove this?

File: coala_quickstart/generation/SettingsFilling.py [hunk 7]
```diff
@@ -165,3 +168,93 @@ def resolve_anomaly(setting_name,
...
+    user_input = input()
+
+    if setting_info["type"]:
+        try:
+            user_input = setting_info["type"](user_input)
+        except ValueError:
+            print(colored(
+                STR_REPORT_INVALID_VALUE_TYPE.format(setting_info["type"]),
+                REPORT_INVALID_TYPE_COLOR))
+            return require_setting(setting_name, setting_info, section)
```
Comment (adtac) [inline coala_quickstart/generation/SettingsFilling.py]: I really hope the user doesn't make the same mistake for a thousand times (999 is the recursion limit in Python IIRC) :P

Comment (adtac) [inline coala_quickstart/generation/SettingsFilling.py]: a `while True` loop would be better here methinks

Comment (adtac) [inline coala_quickstart/generation/SettingsFilling.py]: although this is a clever way to do this too; this is the first time I'm seeing it done this way

File: coala_quickstart/generation/SettingsFilling.py [hunk 8]
```diff
@@ -210,8 +213,20 @@ def require_setting(setting_name, setting_info, section):
     user_input = input()
 
     if setting_info["type"]:
+
         try:
-            user_input = setting_info["type"](user_input)
+            if setting_info["type"] is bool:
+                processed_input = " ".join(re.findall("[a-zA-Z0-9\-]+",
+                                                      user_input.lower()))
```
Comment (adtac) [inline coala_quickstart/generation/SettingsFilling.py]: do we need this? If the user enters an incorrect value, just tell them and ask them to reenter

assuming silently is usually not a great idea :P

**Comment by adtac** [comment]
ack https://github.com/coala/coala-quickstart/pull/164/commits/4d9f1cc568b641106cfd7702dfdefc86dc0e291c

File: coala_quickstart/generation/SettingsFilling.py (hunk 7 above)
Comment (satwikkansal) [inline coala_quickstart/generation/SettingsFilling.py]: Agreed. Using while loop will also save us from extra initializations in every recursive calls :)

File: coala_quickstart/generation/SettingsFilling.py (hunk 8 above)
Comment (satwikkansal) [inline coala_quickstart/generation/SettingsFilling.py]: Right. Simplified it to just remove whitespaces and exclamation symbol.

**Comment by satwikkansal** [comment]
reack 9e02afb

**Comment by adtac** [comment]
ack https://github.com/coala/coala-quickstart/pull/164/commits/ba8ea50a2584a6e6f9e923c106c2020612997e06

**Comment by adtac** [comment]
ack https://github.com/coala/coala-quickstart/pull/164/commits/296c27a3e4f22783fa4f9807a7fe03988ac5406e

**Comment by satwikkansal** [comment]
reack cd7a7dd 485b28b 4802812

**Comment by adtac** [comment]
ack https://github.com/coala/coala-quickstart/pull/164/commits/4802812fc16364b394889ad9df05f2b83a62d1e1

**Comment by adtac** [comment]
@rultor merge

**Pull request merged**